### PR TITLE
QMCPACK timing test for parallel offload.

### PR DESCRIPTION
**Pull request by spophale**
This test has been picked from the Redmine Issue https://openmp-ecp.ornl.gov/redmine/issues/45

### CODE REVIEW

**Comment by spophale** [comment]
Oscar and QMCPACK team needs to look at this test to check if it covers all their requirements before we put it in our repo.

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  const clock_t duration1 = (clock() - start) * 1000 / CLOCKS_PER_SEC / N;
+  
+  if(num_dev > 1){
+    start = clock();
+    for (int i = 0; i < N; ++i) {
+      qmcpack_kernel2(num_dev);
+    }
+    const clock_t duration2 = (clock() - start) * 1000 / CLOCKS_PER_SEC / N;
+    printf("Avg time for parallel offloading to multiple devices: %ld ms\n",duration2);
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/qmcpack_parallel_offload.c]: I suggest we use OMPVV_INFOMSG

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      }
+    }
+  }
+}
+
+int main(int argc, char *argv[]) {
+  int i;
+  const int N = 1000;
+  int num_dev = omp_get_num_devices();
+  //printf("No. of devices available %d.\n", num_dev);
```
Comment (josemonsalve2) [inline tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Remove this comment

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c (hunk 1 above)
Comment (spophale) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Will do that after the test is finalized.

Comment (spophale) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Actually, I don't know if it appropriate as I want the timings to be printed without additional flags.

Comment (josemonsalve2) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: The idea of the TEST_VERBOSE flag was to be able to obtain a simple pass fail report and no other output. If that's the case, then we either want to re-evaluate this design in the infrastructure, or provide a new flag. I don't think it is a good idea to have `printf`s with no other information.

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c (hunk 2 above)
Comment (spophale) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Done.

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c (hunk 1 above)
Comment (spophale) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: This test is a little different from our usual conformance tests. If we see more of these then maybe we should think of a new macro.

Comment (josemonsalve2) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: I agree this is a different to our conformance tests. That's why I am conflicted about adding information that is output in this test (and only in this test) regarding performance, when our focus has always been V&V. If we want to consider performance, we should consider how to change the infrastructure for performance.

Comment (sunitachandra) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: V nice to see an app. test code! I like that we are printing timings!! Probably better to stick with the macros/ infrastructure design/software practice by turning on TEST_VERBOSE and using OMPVV_INFOMSG - that way we are strictly following our rules that we have set and then encourage anyone else to do the same

Comment (spophale) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: INFOMSGs were designed to give additional information for tests, I fear that the timings will get buried in all other messages. That said, current format is for easy testing during discussion phase where they may want to take it for a test drive. I am all in to design a macro that better fits timing reporting.

Comment (josemonsalve2) [reply tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Yes. This is definitely needed. So far we had considered giving timing information for the duration of the whole test, but we have not discussed what happens when we want to give intermediate timing results.

**Comment by spophale** [comment]
Feedback: Add meaningful work inside of the target regions to be safe from dead code elimination.

File: tests/4.5/application_kernels/qmcpack_parallel_offload.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    for (size_t i = 0; i < N; ++i) {
+      for (size_t j = 0; j < N/2 ; ++j) {
+        int tmp_val = N *i*j;
+      }
+    }
+  }
+}
+
+void qmcpack_kernel4(int ndev) {
+  OMPVV_INFOMSG("qmcpack parrallel offloading to different devices with nowait");
```
Comment (jhdavis8) [inline tests/4.5/application_kernels/qmcpack_parallel_offload.c]: Quick spelling fix, parrallel -> parallel.

**Comment by spophale** [comment]
I am closing this PR as this is the only one with timing. If we were to go in the benchmarking direction, more ground up changes would be required to the V&V.

**Pull request closed without merging**